### PR TITLE
Guard install() call in post_migrate signal to prevent runtime errors

### LINKED ISSUE #4169: [Bug]: Unsafe install() call in post_migrate signal may cause runtime errors
### Description

The run_install signal handler in esp/utils/apps.py assumes that every app’s
models module defines an install() function:

    sender.models_module.install()

Since Django’s post_migrate signal runs for all installed apps, this assumption
is not always valid. Some apps may not have a models module or may not define
install(), which can cause runtime errors or unexpected side effects during
application startup.

In my local setup, this resulted in server startup issues and admin registry
mutation during Django initialization.

### Steps to Reproduce

1. Clone the repository and install dependencies
2. Run database migrations
3. Start the development server using `python manage.py runserver`
4. During app initialization, run_install executes for apps that do not define install()

This may trigger runtime errors depending on the app configuration.

### Expected Behavior

The post_migrate handler should only call install() for apps that explicitly
define it in their models module. Server startup and app initialization should
not fail due to missing install() methods.

### Actual Behavior

The signal handler calls sender.models_module.install() unconditionally,
which can cause runtime errors or side effects when the models module does not
define install(). This led to server startup instability in my environment.

### Screenshots

_No response_

### Operating System

_No response_

### Browser

_No response_

### Additional Context

Proposed fix:
Guard the call with a defensive check:
```python
models_module = getattr(sender, "models_module", None)
if models_module and hasattr(models_module, "install"):
    models_module.install()
```

### PR DESCRIPTION
**Pull request by aaditya-kr5892**
## Description

This change adds a defensive check before invoking install() in the
run_install post_migrate signal handler.

The previous implementation assumed that every app’s models module defines an
install() function, which is not always true. Since post_migrate fires for all
installed apps, this could lead to runtime errors or side effects during
application startup.

The fix preserves existing behavior while preventing crashes by ensuring
install() is only called when present.

## Related Issue

Closes #4169

## Type of Change

- [x] Bug fix
- [ ] New feature
- [ ] Breaking change
- [ ] Documentation update
- [ ] Refactor / cleanup
- [ ] CI/CD or build change

## Testing

- [x] Existing tests pass
- [ ] New tests added (if applicable)
- [x] Manually verified

## AI Disclosure

AI assistance was used to help understand the codebase and draft the defensive
fix, but the issue reproduction, debugging, and validation were performed
manually.

## Checklist

- [x] Self-reviewed the code
- [ ] Updated documentation (if needed)
- [x] No new warnings or errors introduced

### CODE REVIEW

**Comment by kkbrum** [comment]
Please address copilot's review when you get a chance.

**Comment by aaditya-kr5892** [comment]
> Please address copilot's review when you get a chance.

working on it

**Comment by willgearty** [comment]
@aaditya-kr5892 any update on this?

**Comment by aaditya-kr5892** [comment]
> @aaditya-kr5892 any update on this?

Hi @willgearty, really sorry for being late in addressing Copilot's review.
I have updated the PR to address all three of Copilot's review comments:
 
- Removed the sys.argv guard: The fix is now strictly focused on the missing install() crash without altering runserver behavior.

- Used callable(): Replaced the hasattr check with getattr + callable() to safely ensure the attribute is an executable function.
 
- Added unit tests: Added RunInstallTestCase to tests.py to cover the missing module, missing install, non-callable install, and happy-path edge cases.

File: esp/esp/utils/tests.py [hunk 1]
```diff
@@ -570,6 +572,42 @@ def test_fast_path_no_data_colon(self):
         self.assertEqual(result, html)
         self.assertEqual(count, 0)
 
+class RunInstallTestCase(unittest.TestCase):
+    """Regression tests for the run_install post_migrate signal handler."""
+
+    def test_sender_with_no_models_module(self):
+        """run_install should not raise when sender.models_module is None."""
+        sender = Mock(spec=[])
+        sender.models_module = None
```
[the remark this answers is not in the document]
Comment (aaditya-kr5892) [reply esp/esp/utils/tests.py]: I just updated the test to test_sender_with_no_models_module using spec_set=[] so it accurately tests the missing attribute case rather than just a None value.

**Review by willgearty — APPROVED** [review]
Looks good, thanks!
